### PR TITLE
fix blockdb deadlock

### PR DESCRIPTION
**Pull request by Greg-Griffith**

### CODE REVIEW

File: src/blockstorage/sequential_files.cpp [hunk 1]
```diff
@@ -154,7 +154,7 @@ uint64_t CalculateCurrentUsage()
 /* Prune a block file (modify associated database entries)*/
 void PruneOneBlockFile(const int fileNumber)
 {
-    READLOCK(cs_mapBlockIndex);
+    RECURSIVEREADLOCK(cs_mapBlockIndex);
     for (BlockMap::iterator it = mapBlockIndex.begin(); it != mapBlockIndex.end(); ++it)
```
Comment (ptschip) [inline src/blockstorage/sequential_files.cpp]: What deadlock are you trying to fix?

Which locks are involved with the deadlock?

Comment (Greg-Griffith) [reply src/blockstorage/sequential_files.cpp]: it is just this lock. it deadlocks itself. this started happening when we switched from a recursive mutex to a shared mutex. some of the blockdb sync code would recurively lock itself which now gets stuck.

I want to test an IBD of dev prior to this PR and then with this PR to see how the rsm performs. is there a huge performance hit or not?

edit: will put WIP on this PR for now

**Comment by gandrewstone** [comment]
cs_mapBlockIndex needs to be the highest performance lock we can get.  Its locked and unlocked all the time.  We need to keep it as a shared lock and find where its being doubly locked in the blockdb code.  Often A calls B and both take the same lock.  In this case, make a lock-free function _B, and change B to { LOCK(theLock); _B(); }  Now A can call _B.  Other times you'll find that A calls B but A doesn't have to hold the lock when B is called so you won't have to make an _B function, just reorganize A.

Yes, there's a beauty, effort, and abstraction penalty we pay for optimization.

**Comment by Greg-Griffith** [comment]
simply moving the SyncStorage() call before cs_mapblockindex is locked in LoadBlockIndexDB() seems to be a situational safe alternative to changing the mutex type. 
There should be no issues since the program has only spawned one thread at this point

**Pull request merged**